skrf/network.py

```diff
@@ -2282,7 +2282,14 @@ def get_buffer() -> io.StringIO:
                 output.write('!Data is not renormalized\n')
                 output.write('# {} S {} R\n'.format(ntwk.frequency.unit, form))
             else:
-                output.write('# {} S {} R {} \n'.format(ntwk.frequency.unit, form, r_ref))
+                # Write "r_ref.real" instead of "r_ref", so we get a real number "a" instead
+                # of a complex number "(a+0j)", which is unsupported by the standard Touchstone
+                # format (non-HFSS). We already checked in the beginning that "r_ref" must be
+                # real in this case (write_z0 == False).
+                assert r_ref.imag == 0, "Complex reference impedance is encountered when " \
```

Comment:
`assert` raises an `AssertionError`, which breaks the program, correct?
I think a `RuntimeWarning` would be better, see this example:
https://github.com/scikit-rf/scikit-rf/blob/21b3409966676d28ac5dfc29552b0eff6c408750/skrf/network.py#L560-L562

Reply:
Oh, the `stacklevel=2` argument is missing in the code above, but that's a good idea as well:
https://docs.python.org/3/library/warnings.html#warnings.warn

Reply:
Yes, it breaks the program, but I believe it's the correct thing to do in this case.

This assertion `r_ref.imag == 0` is guaranteed to be **always true under all conditions**, unless there's another internal bug inside scikit-rf.

## Justification

Complex impedance is not supported to begin with (unless write_z0 is used, which generates a HFSS-style Touchstone file), and in the beginning of write_touchstone(), it actually actively checks whether the imaginary part of the impedance is zero.

```
if r_ref is None and write_z0 == False:
    if not equal_z0:
        raise ValueError("Network has unequal port impedances but reference impedance for renormalization 'r_ref' is not specified.")
    if have_complex_ports:
        raise ValueError("Network port impedances are complex but reference impedance for renormalization 'r_ref' is not specified.")
    r_ref = ntwk.z0[0, 0]
elif r_ref is not None:
    if not npy.isscalar(r_ref):
        raise ValueError('r_ref must be scalar')
    if r_ref.imag != 0:
        raise ValueError('r_ref must be real')
    ntwk.renormalize(r_ref)
```

Thus, r_ref must always be a real number (assert r_ref.imag == 0) when we execute. The complex number output format is purely a result of the string formatting bug as I reported. No complex number support was intended here.

## Summary

In other words, if `r_ref.imag != 0`, scikit-rf already raises a `ValueError` before it hits the "impossible" AssertionError.

If you still think an assertion is inappropriate, please continue the discussion. I don't rule out the possibility that I have missed something, but as far as I know, I don't see any reason not to raise an `AssertionError`.

Reply:
I was ready to write that in `skrf` coding style, we mostly reserve `assert` for test cases scripts (`assert` are removed if python is executed with optimization `-o`, but raised exceptions remains, so assert are like `__debug__` stuffs useful for quickly locating bugs that should not be runtime errors). 

But actually, it appears `assert` is used at several places in the code outside test cases: 

https://github.com/scikit-rf/scikit-rf/blob/21b3409966676d28ac5dfc29552b0eff6c408750/skrf/network.py#L1674

https://github.com/scikit-rf/scikit-rf/blob/7ad561755afbb6c939baa4c68a0d49b96d4c5684/skrf/vi/vna/hp8510c_sweep_plan.py#L59

https://github.com/scikit-rf/scikit-rf/blob/6928886fbb99744e1e77f7843045a2497cc67c99/skrf/qfactor.py#L698

https://github.com/scikit-rf/scikit-rf/blob/1e998a967e523a7cd48cf5755640a0e5b6fe2f18/skrf/mathFunctions.py#L1026

https://github.com/scikit-rf/scikit-rf/blob/f9d0c6f1c5b8d271181ca40e3e892f73a4b599d4/skrf/calibration/calibration.py#L3880

Reply:
Also pinging @Ttl for comment.

Reply:
> I was ready to write that in skrf coding style, we mostly reserve assert for test cases scripts (assert are removed if python is executed with optimization -o, but raised exceptions remains

Good point. If it's the case, would it be better to repeat the same `raise ValueError('r_ref must be real')` at the beginning of the file instead of using an `assert`?

Reply:
In this specific situation (statement that would never be reached during runtime), it would require the function to be modified to trigger the assertion. This would probably happens during development e.g. within a test case and not in a user script.
Both approach seems acceptable to me but I let the other share their mind.

Reply:
Thanks for the detailed comments. I have indeed missed the other previous checks of the realness.
In this case where the warning/error should never get executed, I don't have a strong opinion on how to do it anymore. Throwing another `ValueError` seems reasonable.

Reply:
assert should be fine since it is used to check for assumptions that should always hold.

Reply:
The current consensus seems to be that the use of `assert` is acceptable, and since its only purpose is alerting future developers, it doesn't need to be replaced by a `ValueError`, 

Any additional comment? Is it okay to be merged?